src/Frame/Frame.cc

```diff
@@ -2181,4 +2189,345 @@ void Frame::CloseCachedImage(const std::string& file) {
+        return false;
+    }
+
+    down_sampled_width = std::ceil((float)tile_original_width / mip);
+    down_sampled_height = std::ceil((float)tile_original_height / mip);
+    std::vector<float> tile_data;
+    bool use_loader_downsampled_data(false);
+
+    // Check does the (HDF5) loader has the right (mip) downsampled data
+    if (_loader->HasMip(mip) && _loader->GetDownsampledRasterData(down_sampled_data, channel, stokes, bounds, mip, _image_mutex)) {
+        return true;
+    } else {
+        // Check is there another downsampled data that we can use to downsample
+        for (int sub_mip = 2; sub_mip < mip; ++sub_mip) {
+            if (mip % sub_mip == 0) {
+                int loader_mip = mip / sub_mip;
+                if (_loader->HasMip(loader_mip) &&
+                    _loader->GetDownsampledRasterData(tile_data, channel, stokes, bounds, loader_mip, _image_mutex)) {
+                    use_loader_downsampled_data = true;
+                    // Reset mip
+                    mip = sub_mip;
+                    // Reset the original tile width and height
+                    tile_original_width = std::ceil((float)tile_original_width / loader_mip);
+                    tile_original_height = std::ceil((float)tile_original_height / loader_mip);
+                    break;
+                }
+            }
+        }
+    }
+
+    if (!use_loader_downsampled_data) {
+        // Get full resolution raster tile data
+        int x_min = bounds.x_min();
+        int x_max = bounds.x_max() - 1;
+        int y_min = bounds.y_min();
+        int y_max = bounds.y_max() - 1;
+
+        casacore::Slicer tile_section = GetImageSlicer(AxisRange(x_min, x_max), AxisRange(y_min, y_max), AxisRange(channel), stokes);
+        tile_data.resize(tile_section.length().product());
+        if (!GetSlicerData(tile_section, tile_data.data())) {
+            return false;
+        }
+    }
+
+    // Get down sampled raster tile data by block averaging
+    down_sampled_data.resize(down_sampled_height * down_sampled_width);
+    return BlockSmooth(tile_data.data(), down_sampled_data.data(), tile_original_width, tile_original_height, down_sampled_width,
+        down_sampled_height, 0, 0, mip);
+}
+
+void Frame::GetTiles(int image_width, int image_height, int mip, std::vector<carta::Tile>& tiles) {
+    int tile_size_original = TILE_SIZE * mip;
+    int num_tile_columns = ceil((double)image_width / tile_size_original);
+    int num_tile_rows = ceil((double)image_height / tile_size_original);
+    int32_t tile_layer = carta::Tile::MipToLayer(mip, image_width, image_height, TILE_SIZE, TILE_SIZE);
+    tiles.resize(num_tile_rows * num_tile_columns);
+
+    for (int i = 0; i < num_tile_columns; ++i) {
+        for (int j = 0; j < num_tile_rows; ++j) {
+            tiles[j * num_tile_columns + i].x = i;
+            tiles[j * num_tile_columns + i].y = j;
+            tiles[j * num_tile_columns + i].layer = tile_layer;
+        }
+    }
```

Comment:
One more comment. I am not 100% sure, but I think inverting the loop order will provide better CPU caching for traversing the vector. Now, the double for loop is accessing elements that aren't contiguous en each pass.

I suggest:
```
    for (int j = 0; j < num_tile_rows; ++j) {
        for (int i = 0; i < num_tile_columns; ++i) {
            tiles[j * num_tile_columns + i].x = i;
            tiles[j * num_tile_columns + i].y = j;
            tiles[j * num_tile_columns + i].layer = tile_layer;
        }
    }
```